aten/src/ATen/native/nested/README.md

```diff
@@ -0,0 +1,63 @@
+- `NestedTensorFactories.h | NestedTensorFactories.cpp`: Functions for creating NestedTensors (e.g. empty_like)
+- `NestedTensorMath.h | NestedTensorMath.cpp`: Math functions on NestedTensors (e.g. softmax, embedding)
+- `NestedTensorMatmul.cpp`: Matmul functions on NestedTensors (e.g. matmul, linear, bmm)
+- `NestedTensorTransformerFunctions.h | NestedTensorTransformerFunctions.cpp`: Functions for enabling the BetterTransformer work stream
+- `cuda/`: CUDA implementations of the NestedTensor functions
+
+##  Implementing new functions
+There are two main classes of functions that can be implemented on NestedTensors: functions that can be efficiently implemented by viewing the NestedTensor as dense tensor where the raggedness has been folded into the dense dimensions and those that can't. Unary operations (e.g. abs, log, relu) and binary operations (e.g. add, mul, div) that act elementwise on one or more NestedTensors are examples of functions that can be efficiently implemented. Efficient implementation of these functions is relatively straightforward.
+
+The definition of map_nt is:
+
+```cpp
+template <typename Func>
+Tensor map_nt(const Tensor& nt, Func f) {
+  auto* nt_impl = get_nested_tensor_impl(nt);
+  const auto& sizes = nt_impl->get_nested_size_tensor();
+  return at::detail::make_tensor<NestedTensorImpl>(f(nt_impl->get_buffer()), sizes);
+}
+```
+1. Get the NestedTensorImpl from the input NestedTensor.
+2. Get the sizes of the NestedTensor.
+3. Call get_buffer() which returns a flat, dense tensor whose storage shares that of the input NestedTensor.
+4. Call the function f on the dense tensor.
+5. Construct a new NestedTensor from the output of f and the sizes of the input NestedTensor.
+
+There are also important functions that, under certain conditions of regularity, can be implemented effectively by accessing the underlying buffer and viewing it in a special manner. For a good example of this see the implementation of `linear` in [NestedTensorTransformerFunctions.cpp](NestedTensorTransformerFunctions.cpp).
+
+The second class of functions can't be efficiently implemented by viewing the NestedTensor as a dense tensor. An example of this is `softmax_nested`. The implementation of this function can be found in [NestedTensorMath.cpp](NestedTensorMath.cpp). This can occur when there are strict problem boundaries that occur at the tensor_component level. For example, when computing the softmax of a NestedTensor we need to compute the softmax of each tensor_component individually. This is because the softmax function is not defined on tensors of different shapes. In this case we can't view the NestedTensor as a dense tensor and apply the softmax function to the entire tensor at once. Instead we need to apply the softmax function to each tensor_component individually.
```

Comment:
I don't think this is necessarily true if the last dimension of the NestedTensor is regular (same for any other normalization)? [softmax](https://pytorch.org/docs/master/generated/torch.nn.functional.softmax.html?highlight=softmax#torch.nn.functional.softmax) is applied at the per-vector level. The reason it's an issue for us right now, is because we apply softmax as part of SDPA's math decomposition where the matmul between query and key cause irregular last dimensions.

Reply:
ahhh you are right I think I let co-pilot get a little carried away. Do you think it would be fair to say:
```
The second class of functions can't be efficiently implemented by viewing the NestedTensor as a dense tensor. An example of this is `softmax_nested` of ragged dimensions. The implementation of this function can be found in [NestedTensorMath.cpp](NestedTensorMath.cpp). When computing the softmax over a ragged dimension of a NestedTensor the problem boundaries are not trivially separable, i.e. it is not possible to determine which elements belong to what tensor_component when folding the ragged dimension into another. Instead we apply the softmax function to each tensor_component individually, essentially applying a foreach loop over the individual tensors in the NestedTensor.
```